Join stake pool endpoint (part 2 - add sign delegation cert)

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#896 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have adjusted naming as suggested by Johannes in part 1 PR
- [x] I have added signDelegationCert impl
- [x] I have shared create/sign/submit between Tx and DelegationCert 
- [x] I have adjusted and added new integration tests

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
rename stuff as suggested by Johannes in previous PR

paweljakubas committed:
add signDelegationCert impl plus error cstr handling from Transaction layer

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -112,12 +112,12 @@ module Cardano.Wallet
+    , submitDelegationCert
+    , ErrCreateDelegationCert (..)
+    , ErrSignDelegationCert (..)
+    , ErrSubmitDelegationCert (..)
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1021,23 +1024,71 @@ signCert
+            let time = slotStartTime
+                    (slotParams (blockchainParameters cp))
+                    txSlot
+            return (undefined, meta, time, undefined)
```
**[question]**

What about trying to factor the common parts of this and `signTx` in one or more composable functions? I see arguably at least 2 functions:

- change creation

```hs
            let (changeOuts, s') = flip runState (getState cp) $
                    forM chgs $ \c -> do
                        addr <- state (genChange argGenChange)
                        return $ TxOut addr c
            let allOuts = outs ++ changeOuts
```

- metadata construction

```hs
            let ourCoins (TxOut addr (Coin val)) =
                    if fst (isOurs addr s')
                        then Just (fromIntegral val)
                        else Nothing
            let amtOuts = sum (mapMaybe ourCoins allOuts)
            let amtInps = fromIntegral $
                    sum (getCoin . coin . snd <$> ins)
            let txSlot =
                    (currentTip cp) ^. #slotId
            let meta = TxMeta
                    { status = Pending
                    , direction = Outgoing
                    , slotId = txSlot
                    , blockHeight = (currentTip cp) ^. #blockHeight
                    , amount = Quantity (amtInps - amtOuts)
                    }
            let time = slotStartTime
                    (slotParams (blockchainParameters cp))
                    txSlot
```

Perhaps making them separate function internal to the module would help to identify further commonalities?

KtorZ replied:
done.

KtorZ committed:
re-locate 'signDelegationCert' and 'createDelegationCert' closer to their standard counterparts

Help reducing possible discrepency when updating code and, help also
visualizing possible abstractions

KtorZ committed:
rename public module API regarding transaction creations

A pattern seems to have emerged about 'selecting coins', then signing the selection and finally, submitting the
transaction. So, I've tried to use names that better reflect this patterns and somewhat clarifies a bit how these
pieces work together

KtorZ committed:
define 'mkTxMeta' to factor out metadata creation between payment and delegation

KtorZ committed:
add 'deriveRewardAccount' method to 'AddressDerivation' instead of having this logic inside the wallet engine

KtorZ committed:
factor out gen change commonalitites into a 'assignChangeAddresses' function

KtorZ committed:
implement remaining logic for delegation certificate submission using mkDelegationCertTx

KtorZ committed:
revise error handling and message of 'ErrKeyNotFound...' (actually a 500)

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -820,22 +819,22 @@ joinStakePool ctx spl (ApiT poolId) (ApiT wid) passwd = do
 
     (tx, meta, time, wit) <- liftHandler $ withWorkerCtx ctx wid liftE2 $ \wrk ->
-        W.signCert @_ @s @t @k wrk wid () pwd selection poolId
+        W.signDelegation @_ @s @t @k wrk wid () pwd selection poolId
```
I'm wondering if we don't have a potential race-condition with `POST /transaction` and `joinStakePool`:

We run:
1. `selectCoinsForDelegation` does the coin selection based on the `availableUtxo`.
2. `signDelegation`
3. `submitTx` — modifies the `availableUtxo` by adding a pending transaction

Step 1 and 3 are run in different DB transactions.

So if we in parallel run:
```
joinStakePool            1           23
joinStakePool                  1       2   3
-------------------------------------------------> time
```
we could end up (attempting to) spend the same funds twice.

KtorZ replied:
Indeed, but that's nothing new :) 
We already have that particular race condition from making two concurrent payments in parallel. Worst that can happen is that we create and sign both transactions, we submit both, and only one get through in the end while the other stay pending forever.

KtorZ replied:
> we could end up spending the same funds twice.

No! The network will prevent that!

Anviking replied:
> No! The network will prevent that!

🙄 Ok, bad wording on my part

KtorZ committed:
fix output reference in Api server for join stake pool

The selection's output is actually empty before balancing! So we need to have a look at the the transaction outputs

KtorZ committed:
enable stake pools delegation integration scenarios

Co-authored-by: Pawel Jakubas <pawel.jakubas@iohk.io>

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -811,49 +870,118 @@ signTx
+    => BlockchainParameters
+    -> BlockHeader
+    -> s
+    -> [(TxIn, TxOut)]
```
(unrelated) It's been nagging me for a while, but we should maybe add proper types for `(TxIn, Coin)` and `(TxIn, TxOut)` sometime

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -811,49 +870,118 @@ signTx
+            let rewardAccount = deriveRewardAccount @k xprv pwd
+
+            let hrp = Bech32.unsafeHumanReadablePartFromText "addr"
+            let accountAddr
```
Is importing `Bech32` and knowing how to create the reward `Address` in `core` is temporary? (it is target-specific)

KtorZ replied:
![Screenshot from 2019-11-27 19-17-39](https://user-images.githubusercontent.com/5680256/69749358-a1525100-114a-11ea-8d1d-c040edad11a6.png)

It is, as indicated by this big fat "TEMPORARY" in the corresponding commit :sweat_smile:

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -305,6 +308,20 @@ class HardDerivation key => SoftDerivation (key :: Depth -> * -> *) where
+    -> k 'AddressK XPrv
+deriveRewardAccount rootPrv pwd =
+    let accPrv = deriveAccountPrivateKey pwd rootPrv minBound
+    in deriveAddressPrivateKey pwd accPrv MutableAccount minBound
```
👍

Anviking commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -128,11 +134,26 @@ spec = do
+            request @[ApiTransaction n] ctx ep Default Empty >>= flip verify
+                [ expectListItemFieldEqual 0 direction Outgoing
+                , expectListItemFieldEqual 0 status InLedger
+                ]
```
👍 

Are we eventually going to test that rewards accumulate in the reward account too?

Anviking replied:
Hm, that must depend on #904. Maybe I should pick it up.

paweljakubas replied:
yes, requesting joining the stake pool is the first step. the next one is to discover it when accommodated in blockchain

paweljakubas committed:
add cannot rejoin the same stake pool test

KtorZ committed:
add 'deriveRewardAccount' method to 'AddressDerivation'

To avoid writing this manually in a few different places

KtorZ committed:
re-locate 'liftPaymentFingerprint' inside 'DelegationAddress'

We will only be generating delegation addresses (i.e. 'grouped') and we therefore need
to lift from a payment fingerprint into a grouped address

KtorZ committed:
remove 'delegationKeyFingerprint' from 'MkKeyFingerprint'

It's not needed and it makes implementing other more interesting feature more complicated

KtorZ committed:
Allow 'MkKeyFingerprint' to be used from different sources (address or public key)

This allows to transform the 'PaymentAddress' constraints in the discovery module into less powerful
'MkKeyFingerprint' constraints, and remove altogether the network type parameter on address pools.

KtorZ committed:
re-introduce the ability to lift fingerprints into payment addresses

This will be particularly useful to avoid storing too large addresses in the database.
Ideally, we should store only the fingerprints into the database, but
this requires even more work so for now, we are going through the
following transformations:

- Address pool keeps track of 'fingerprints'
- When stored inside the database, we lift the fingerprints to payment addresses
- Then, when read from the database, we extract the fingerprints from the addresses again

KtorZ committed:
allow storing the reward account public to the database

KtorZ committed:
only use 'DelegationAddress' in the API

KtorZ committed:
update remaining spec files to use delegation addresses only

KtorZ committed:
normalize tx history out addrs to delegated addrs when listing them

KtorZ reviewed (approved):
@paweljakubas 

- Rebased on top of `KtorZ/896/delegation-addresses-only`
- Removed the temporary hack
- Test now pass as expected :tada:

KtorZ committed:
revise rewardAccount construction to use top-level API already available for key derivation
